test/matrix_utils.hh

```diff
@@ -203,4 +205,146 @@ matrix_type matrix_cast(
     return matrix_cast_impl( MatrixType< matrix_type >(), A, uplo, diag );
 }
 
+
+// -----------------------------------------------------------------------------
+// Functions for allocating test matrices
+
+template <typename MatrixType>
+class TestMatrix {
+    using scalar_t = typename MatrixType::value_type;
+
+public:
+    // SLATE matrices
+    MatrixType A;
+    MatrixType Aref;
+
+    // Storage for ScaLAPACK matrices
+    std::vector<scalar_t> A_data;
+    std::vector<scalar_t> Aref_data;
+
+    // ScaLAPACK configuration
+    int64_t m, n, mloc, nloc, lld, nb;
+
+    #ifdef SLATE_HAVE_SCALAPACK
+    void ScaLAPACK_descriptor( blas_int ictxt, blas_int A_desc[9] ) {
+        int64_t info;
+        scalapack_descinit(A_desc, m, n, nb, nb, 0, 0, ictxt, mloc, &info);
+        slate_assert(info == 0);
+    }
+    #endif
+};
+
+// -----------------------------------------------------------------------------
+/// Marks the paramters used by allocate_test_Matrix
+inline void mark_params_for_test_Matrix(Params& params)
+{
+    params.grid.m();
+    params.grid.n();
+    params.nb();
+    params.nonuniform_nb();
+    params.origin();
+    params.grid_order();
+}
+
+// -----------------------------------------------------------------------------
+/// Allocates a Matrix<scalar_t> and a reference version for testing.
+///
+/// @param ref_matrix[in]
+///     Whether to allocate a reference matrix
+///
+/// @param nonuniform_ref[in]
+///     If params.nonuniform_nb(), whether to also allocate the reference matrix
+///     with non-uniform tiles.
+///
+/// @param m[in]
+///     The number of rows
+///
+/// @param n[in]
+///     The number of columns
+///
+/// @param params[in]
+///     The test params object which contains many of the key parameters
+///
+template <typename scalar_t>
+TestMatrix<slate::Matrix<scalar_t>> allocate_test_Matrix(
+        bool ref_matrix,
+        bool nonuniform_ref,
+        int64_t m,
+        int64_t n,
+        Params& params)
```

Comment:
You're right that this is really just a constructor.  However, in the follow up PR (#157), I add `allocate_test_HermitianMatrix` et al.  So, if we did these functions as constructors, we'd need to conditionally disable them with `std::enable_if`.  I'd be inclined to at least leave it until #157 is considered, so that they can all be done consistently.